packages/controllers/src/plugins/PluginController.ts

```diff
@@ -235,17 +240,38 @@ export class PluginController extends BaseController<
     this._terminateAllPlugins = terminateAllPlugins;
     this._executePlugin = executePlugin;
     this._getRpcMessageHandler = getRpcMessageHandler;
+    this._onUnhandledPluginError = this._onUnhandledPluginError.bind(this);
+    this._onUnresponsivePlugin = this._onUnresponsivePlugin.bind(this);
+
     this.messagingSystem.subscribe(
       'ServiceMessenger:unhandledError',
-      (pluginName, error) => {
-        this.stopPlugin(pluginName);
-        this.addPluginError(error);
-      },
+      this._onUnhandledPluginError,
+    );
+
+    this.messagingSystem.subscribe(
+      'ServiceMessenger:unresponsive',
+      this._onUnresponsivePlugin,
     );
 
     this._pluginsBeingAdded = new Map();
   }
 
+  _onUnresponsivePlugin(pluginName: string) {
+    this.stopPlugin(pluginName);
+    this.addPluginError({
+      code: -32001, // just made this code up
+      message: 'Plugin Unresponsive',
+      data: {
+        pluginName,
+      },
+    });
+  }
```

Comment:
Follow-up items (for a future PR):

- We should consider making the default behavior of stopping unresponsive plugins configurable.
- We should specify an error code for this, and then use an enum for the value. We need to find a good, unspoken-for integer range we can use for snap errors.